### PR TITLE
Unicode EN (half an EM) support

### PR DESCRIPTION
**Pull request by ghost**
In Unicode, there are only three characters whose display width should be one cell. (EN QUAD / EN SPACE / EN DASH).
It is a problem with CJKV.
ambiwidth should not be applied to these characters.
When it is applied, it looks like this.

image

This screen is written in Japanese.
(EN DASH) is used. It is a short horizontal bar.
In this example, ambiwidth is 2.
In ConPTY at Vim, the rows are swapped.
In winpty in Vim, the rightmost character is not output.
In cmd.exe, it will display normally.

When this patch is applied, the cell width of (EN DASH) becomes 1 and it is displayed normally.

### CODE REVIEW

**Comment by ghost** [comment]
(EN QUAD) is C/D normalized (EN SPACE).
(EN SPACE) is KC/KD normalized to U+0020.
Both are U+0020, which is a blank space of cell width 1.

**Comment by ghost** [comment]
After applying the patch.
(EN DASH) with a red underline.

image

File: src/mbyte.c [hunk 1]
```diff
@@ -1414,6 +1414,15 @@ utf_uint2cells(UINT32_T c)
...
 utf_char2cells(int c)
 {
+    /* Sorted list of non-overlapping intervals of EN(single width characters
+     * in typesetting) that are not affected by p_ambw and CJKV. */
+    static struct interval singlewidth[] =
+    {
+	{0x2000, 0x2000},
+	{0x2002, 0x2002},
+	{0x2013, 0x2013}
+    };
```
Comment (mattn) [inline src/mbyte.c:1424]: I don't understand why you make this list. Just removing the entries from variable `ambiguous`? Also I noticed `\u2013` is defined as `A` in http://www.unicode.org/Public/11.0.0/ucd/EastAsianWidth.txt

Comment (mattn) [reply src/mbyte.c:1424]: `\u2000` and `\u2002` both not exists in doublewidth/emoji_width/ambiguous.

Comment (ghost) [reply src/mbyte.c:1424]: I listed the letters with EN attached.  It was a coarse patch. I will review again.

**Comment by ghost** [comment]
\u2013 for special processing. Otherwise, it can not be compatible with cmd.exe.
An indication from @mattn, mintty draw 2 cells as ambiguous width.
I will think further.
I thought about now that we can set the cell width of  \u2013 in the variable 'enwidth'.
Or special processing by detecting cmd.exe.
Otherwise it will allow destruction of the display.

I feel like I should not do anything...

**Comment by mattn** [comment]
I wrote some ideas in this comment: https://github.com/vim/vim/pull/2074#issuecomment-328699857

**Comment by ghost** [comment]
I think that it is good design to read the text file which defined the range.
Since I write a demonstration sample, I want to advance the topic.
If this is not over, I will not be able to proceed with ConPTY because of the display.

**Comment by ghost** [comment]
My act is to make a roof on the roof.
I do other things to do.
I need more thought.

**Comment by ghost** [comment]
I started with the detection of cmd.exe.
Cause of the screen being incorrect at the compatible terminal, now there is also a patch for it, this will be divided into different patches.
I keep ConEmu's detection in, so I will be a little slovenly, I will be careful.

**Comment by ghost** [comment]
I think that it was able to correspond to the whole query.
It would be cool if this was available from the beginning.

The width of EN DASH is Japanese-specific. The value of this patch itself is questioned.
EN DASH becomes one cell only when cmd.exe is Japanese. There are infinite combinations.
It is useful if you can use queries, think about other uses.

Please ignore this issue.
At a later date, reapply as a patch with another purpose and link.

**Comment by chrisbra** [comment]
so closing.

**Comment by mattn** [comment]
Ref: https://github.com/vim/vim/commit/08aac3c6192f0103cb87e280270a32b50e653be1

**Pull request closed without merging**